fix: do not require rules but instead pass them directly as objects

dword-design opened the pull request.

redonkulus reviewed (commented):
Is this a breaking change?

tom76kimo commented:
@redonkulus Looks like one. I will rebuild the pipeline and merge if it passes.

dword-design commented:
@redonkulus @tom76kimo Yeah I think so.

AntonioRedondo commented:
These three @dword-design's pull requests seem interesting. No chance for any of them to be reviewed and merged?

redonkulus commented:
@dword-design can we maintain backwards compat and support both options? With preference for the rules object being passed in, then falling back to the file.

dword-design commented:
@redonkulus I've just made it backward compatible on my branch. If `rules` is a string, it's loaded from file, otherwise it's taken as-is. For some reason the changes aren't visible in the PR. You can check out my branch here:
https://github.com/dword-design/webpack-atomizer-loader/commits/2020-02/rules-as-object

redonkulus commented:
Changes seem fine to me, want to push them to this branch so I can see them here?

dword-design commented:
@redonkulus They are on the PR branch, don't know why they aren't visible here.

redonkulus commented:
Github was having issues, possibly its related. You sure `dword-design:2020-02/rules-as-object` is updated?

dword-design commented:
@redonkulus Yeah on the linked branch the commit is there and here it's not. I had GitHub-related issues pushing the changes, but they are there now and when checking out the branch they are also there.

dword-design commented:
@redonkulus Alright I triggered an empty commit and now it's there.

redonkulus commented on lib/atomicLoader.ts:
```diff
@@ -52,6 +52,16 @@ const parseAndGenerateFile = function(
 
         const pathConfig: PathConfig = configObject[configPath] || configObject.default;
+
+        // custom rules file
```
Curious why this code block was moved above the `atomizer.findClassNames` statement?

dword-design replied:
I tested it in the old order and actually there some issues here.

Firstly, the rules are added after the classes are found via Atomizer. This does not work because Atomizer won't find classes that belong to newly added rules.

Secondly, rules are added to a global Atomizer object, which actually isn't too good because multiple instances of atomic-loader will share rules (which does not sound intended). In production not really a problem but in testing it is because you want to test multiple atomic-loaders. That's why interestingly only one test fails, the other succeeds since the rule from the previous one is still there ^^.

I'm actually not sure if custom rules have ever worked, also because there are no tests about them.

redonkulus commented on lib/atomicLoader.ts:
```diff
@@ -52,6 +52,16 @@ const parseAndGenerateFile = function(
+        if (pathConfig.options && pathConfig.options.rules) {
+            atomizer.addRules(
+                typeof pathConfig.options.rules === 'string'
+                    ? require(require.resolve(pathConfig.options.rules))
```
previous code guarded against teh require and prevented the `atomizer.addRules` method from being called, do you want to keep this conditional?

dword-design replied:
Not sure about this. I've kept it like it is. It depends on how flexible `atomizer.addRules` is.

redonkulus commented on tests/atomicLoader.js:
```diff
@@ -2,6 +2,7 @@
 var fs = require('fs');
 var expect = require('chai').expect;
 var atomicLoader = require('../dist/atomicLoader');
+var path = require('path')
```
nit: add ';'

dword-design replied:
Done

redonkulus commented:
@dword-design Sorry I broke your PR because I updated all the deps, ran prettier and cleaned up the code a little. Do you mind rebasing again?

redonkulus reviewed (commented):
Do you mind updating the readme to mention passing in rules now?

dword-design commented:
> @dword-design Sorry I broke your PR because I updated all the deps, ran prettier and cleaned up the code a little. Do you mind rebasing again?

Done

> Do you mind updating the readme to mention passing in rules now?

Looks a bit like rules passing is not documented yet. So would be a separate step. I could still do it but I need advice on where to put it. Tbh I'm always a bit lost on the config system in the Atomizer project ^^.

redonkulus commented:
I just meant to document this new config option with a code example of its usage. I haven't read the readme docs thoroughly, if there isn't a place to document the options, I can do it in a follow up PR.

redonkulus commented:
Published in https://github.com/acss-io/webpack-atomizer-loader/releases/tag/v1.2.0
